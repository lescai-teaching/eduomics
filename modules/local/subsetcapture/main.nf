process SUBSETCAPTURE {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/bedtools_htslib:46c2f90c5e5c04ad':
        'community.wave.seqera.io/library/bedtools_htslib:d780123314824c66' }"

    input:
    tuple val(meta), val(chromosome)
    path target_chrom_size
    path capture_bed

    output:
    //tuple val(meta), path("*_target.chrom.sizes"), emit: target_chrom_size
    tuple val(meta), path("*.capture.bed.gz"), emit: capture_bed_gz
    tuple val(meta), path("*.capture.bed.gz.tbi"), emit: capture_bed_index
    tuple val(meta), path("*_target.bed"), emit: target_bed
    tuple val(meta), path("*_target.pad50.bed"), emit: target_bed_pad50
    tuple val(meta), path("*_target.pad500.bed"), emit: target_bed_pad500
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    # Compress and index origin capture BED
    bgzip -c ${capture_bed} > ${prefix}.capture.bed.gz
    tabix -p bed ${prefix}.capture.bed.gz

    # Subset by chrom
    tabix -p bed ${prefix}.capture.bed.gz ${chromosome} > ${prefix}_${chromosome}_target.bed


    #padding
    bedtools slop -i ${prefix}_${chromosome}_target.bed -g ${target_chrom_size} -b 500 > ${prefix}_${chromosome}_target.pad500.bed
    bedtools slop -i ${prefix}_${chromosome}_target.bed -g ${target_chrom_size} -b 50 > ${prefix}_${chromosome}_target.pad50.bed

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        subsetcapture: \$(samtools --version |& sed '1!d ; s/samtools //')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        subsetcapture: \$(samtools --version |& sed '1!d ; s/samtools //')
    END_VERSIONS
    """
}
