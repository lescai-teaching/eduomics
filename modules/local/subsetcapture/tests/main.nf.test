nextflow_process {
    name "Test SUBSETCAPTURE"
    script "../main.nf"
    process "SUBSETCAPTURE"

    test("subset capture - bed") {
        when {
            process {
                """
                input[0] = [ [ id:'test', single_end:false ], 'chr22' ]
                input[1] = file("https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/chrom.sizes")
                input[2] = file("https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/Twist_Exome_Target2_chr22q11-21_GRCh38.bed")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
            assert path(process.out.capture_bed_gz.get(0).get(1)).exists()
            assert path(process.out.capture_bed_index.get(0).get(1)).exists()
            assert path(process.out.target_bed_pad50.get(0).get(1)).exists()
            assert path(process.out.target_bed_pad500.get(0).get(1)).exists()
            //{ assert process.out.reads.get(0).get(1).first().endsWith('.fasta.gz') },
            assert path(process.out.versions.get(0)).exists()
        }
    }
}


