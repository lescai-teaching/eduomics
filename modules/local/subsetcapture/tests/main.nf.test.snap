{
    "subset capture - bed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_target.chrom.sizes:md5,3b8cd381d92f11c1a6cd64a01a5c4b66"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.capture.bed.gz:md5,1bfb2a02b6882800dfc86f1dcd116dea"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.capture.bed.gz.tbi:md5,cca8582edb548b04e057329d3c8de585"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_chr22_target.bed:md5,1bfb2a02b6882800dfc86f1dcd116dea"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_chr22_target.pad50.bed:md5,ac358cce78980c94f6ff5be56cbdede9"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_chr22_target.pad500.bed:md5,4188712633feb7137bc5d29a3940aa6a"
                    ]
                ],
                "6": [
                    "versions.yml:md5,316999bed19877d0178cd728c162e3ca"
                ],
                "capture_bed_gz": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.capture.bed.gz:md5,1bfb2a02b6882800dfc86f1dcd116dea"
                    ]
                ],
                "capture_bed_index": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.capture.bed.gz.tbi:md5,cca8582edb548b04e057329d3c8de585"
                    ]
                ],
                "target_bed": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_chr22_target.bed:md5,1bfb2a02b6882800dfc86f1dcd116dea"
                    ]
                ],
                "target_bed_pad50": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_chr22_target.pad50.bed:md5,ac358cce78980c94f6ff5be56cbdede9"
                    ]
                ],
                "target_bed_pad500": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_chr22_target.pad500.bed:md5,4188712633feb7137bc5d29a3940aa6a"
                    ]
                ],
                "target_chrom_size": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_target.chrom.sizes:md5,3b8cd381d92f11c1a6cd64a01a5c4b66"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,316999bed19877d0178cd728c162e3ca"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-25T14:28:12.505219179"
    }
}